### ISSUE TITLE
1.6.0-RC1 assert: rk->rk_consumer.wait_commit_cnt > 0

### ISSUE THREAD

**Issue by filimonov** [edenhill/librdkafka#3217]
After the attempt of upgrading librdkafka to 1.6.0-RC1 in ClickHouse, one  of the tests start failing with assert from librdkafka. 

(Test checks the behaviour when the connection to brokers lost just before the commit)

```
*** ../contrib/librdkafka/src/rdkafka_cgrp.c:2824:rd_kafka_cgrp_op_handle_OffsetCommit: assert: rk->rk_consumer.wait_commit_cnt > 0 ***
```

```
2021.01.16 01:40:18.925129 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPOP] [thrd:main]: Group "duplicates_when_commit_failed" received op GET_ASSIGNMENT in state up (join-state steady)
2021.01.16 01:40:18.925378 [ 29 ] {} <Trace> StorageKafka (kafka): Polled offset 21 (topic: duplicates_when_commit_failed, partition: 0)
2021.01.16 01:40:18.925554 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPOP] [thrd:main]: Group "duplicates_when_commit_failed" received op OFFSET_COMMIT in state up (join-state steady)
2021.01.16 01:40:18.925705 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:OFFSET] [thrd:main]: Topic duplicates_when_commit_failed [0]: stored offset 21, committed offset 1: setting stored offset 21 for commit
2021.01.16 01:40:18.925805 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:COMMIT] [thrd:main]: GroupCoordinator/1: Committing offsets for 1 partition(s): manual
2021.01.16 01:40:18.925921 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:OFFSET] [thrd:main]: GroupCoordinator/1: Enqueue OffsetCommitRequest(v6, 1/1 partition(s))): manual
2021.01.16 01:40:18.926191 [ 55 ] {} <Debug> StorageKafka (kafka): [rdk:SEND] [thrd:GroupCoordinator]: GroupCoordinator/1: Sent OffsetCommitRequest (v6, 205 bytes @ 0, CorrId 14)
2021.01.16 01:40:20.933675 [ 56 ] {} <Warning> StorageKafka (kafka): [rdk:SESSTMOUT] [thrd:main]: Consumer group session timed out (in join-state steady) after 10497 ms without a successful response from the group coordinator (broker 1, last error was Success): revoking assignment and rejoining group
2021.01.16 01:40:20.933862 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:MEMBERID] [thrd:main]: Group "duplicates_when_commit_failed": updating member id "ClickHouse-instance-test-kafka-5486b636-446f-4816-a95d-291c8322518a" -> ""
2021.01.16 01:40:20.934008 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:REBALANCE] [thrd:main]: Group "duplicates_when_commit_failed" initiating rebalance (EAGER) in state up (join-state steady) with 1 assigned partition(s) (lost): Consumer group session timed out (in join-state steady) after 10497 ms without a successful response from the group coordinator (broker 1, last error was Success)
2021.01.16 01:40:20.934145 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:LOST] [thrd:main]: Group "duplicates_when_commit_failed": current assignment of 1 partition(s) lost: Consumer group session timed out (in join-state steady) after 10497 ms without a successful response from the group coordinator (broker 1, last error was Success): revoking assignment and rejoining
2021.01.16 01:40:20.934254 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPJOINSTATE] [thrd:main]: Group "duplicates_when_commit_failed" changed join state steady -> wait-unassign-call (state up)
2021.01.16 01:40:20.934382 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:ASSIGN] [thrd:main]: Group "duplicates_when_commit_failed": delegating revoke of 1 partition(s) to application on queue rd_kafka_cgrp_new: Consumer group session timed out (in join-state steady) after 10497 ms without a successful response from the group coordinator (broker 1, last error was Success)
2021.01.16 01:40:20.934480 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:PAUSE] [thrd:main]: Pausing fetchers for 1 assigned partition(s): rebalance
2021.01.16 01:40:20.934569 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:PAUSE] [thrd:main]: Library pausing 1 partition(s)
2021.01.16 01:40:20.934662 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:BARRIER] [thrd:main]: duplicates_when_commit_failed [0]: rd_kafka_toppar_op_pause_resume:2415: new version barrier v5
2021.01.16 01:40:20.934744 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:PAUSE] [thrd:main]: Pause duplicates_when_commit_failed [0] (v5)
2021.01.16 01:40:20.934846 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:ASSIGNMENT] [thrd:main]: Group "duplicates_when_commit_failed": clearing group assignment
2021.01.16 01:40:20.934959 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:OP] [thrd:main]: duplicates_when_commit_failed [0] received op PAUSE (v5) in fetch-state active (opv4)
2021.01.16 01:40:20.935051 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:PAUSE] [thrd:main]: Pause duplicates_when_commit_failed [0]: at offset 21 (state active, v5)
2021.01.16 01:40:21.428486 [ 58 ] {} <Debug> StorageKafka (kafka): [rdk:FETCHDEC] [thrd:kafka1:19092/bootstrap]: Topic duplicates_when_commit_failed [0]: fetch decide: updating to version 5 (was 4) at offset 21 (was 23)
2021.01.16 01:40:21.428644 [ 58 ] {} <Debug> StorageKafka (kafka): [rdk:FETCHADD] [thrd:kafka1:19092/bootstrap]: kafka1:19092/1: Removed duplicates_when_commit_failed [0] from fetch list (0 entries, opv 5): paused
2021.01.16 01:40:23.436215 [ 55 ] {} <Notice> StorageKafka (kafka): [rdk:REQTMOUT] [thrd:GroupCoordinator]: GroupCoordinator/1: Timed out HeartbeatRequest in flight (after 10001ms, timeout #0)
2021.01.16 01:40:23.436511 [ 55 ] {} <Warning> StorageKafka (kafka): [rdk:REQTMOUT] [thrd:GroupCoordinator]: GroupCoordinator/1: Timed out 1 in-flight, 0 retry-queued, 0 out-queue, 0 partially-sent requests
2021.01.16 01:40:23.436581 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:HEARTBEAT] [thrd:main]: Group "duplicates_when_commit_failed" heartbeat error response in state up (join-state wait-unassign-call, 0 partition(s) assigned): Local: Timed out
2021.01.16 01:40:23.436721 [ 55 ] {} <Error> StorageKafka (kafka): [rdk:FAIL] [thrd:GroupCoordinator]: GroupCoordinator: kafka1:19092: 1 request(s) timed out: disconnect (after 19022ms in state UP)
2021.01.16 01:40:23.436940 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPSTATE] [thrd:main]: Group "duplicates_when_commit_failed" changed state up -> query-coord (join-state wait-unassign-call)
2021.01.16 01:40:23.437089 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPQUERY] [thrd:main]: kafka1:19092/1: Group "duplicates_when_commit_failed": querying for coordinator: intervaled in state query-coord
2021.01.16 01:40:23.437150 [ 58 ] {} <Debug> StorageKafka (kafka): [rdk:SEND] [thrd:kafka1:19092/bootstrap]: kafka1:19092/1: Sent MetadataRequest (v4, 80 bytes @ 0, CorrId 40)
2021.01.16 01:40:23.437227 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPSTATE] [thrd:main]: Group "duplicates_when_commit_failed" changed state query-coord -> wait-coord (join-state wait-unassign-call)
2021.01.16 01:40:23.437346 [ 58 ] {} <Debug> StorageKafka (kafka): [rdk:SEND] [thrd:kafka1:19092/bootstrap]: kafka1:19092/1: Sent FindCoordinatorRequest (v2, 76 bytes @ 0, CorrId 41)
2021.01.16 01:40:23.437392 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:COMMIT] [thrd:main]: GroupCoordinator/1: OffsetCommit for 1 partition(s) in join-state wait-unassign-call: manual: returned: Local: Operation in progress
2021.01.16 01:40:23.437691 [ 55 ] {} <Debug> StorageKafka (kafka): [rdk:CONNECTED] [thrd:GroupCoordinator]: GroupCoordinator/1: Connected (#2)
2021.01.16 01:40:23.437835 [ 55 ] {} <Debug> StorageKafka (kafka): [rdk:RETRY] [thrd:GroupCoordinator]: GroupCoordinator/1: Retrying OffsetCommitRequest (v6, 205 bytes, retry 1/2, prev CorrId 14) in 100ms
2021.01.16 01:40:23.437991 [ 55 ] {} <Debug> StorageKafka (kafka): [rdk:SEND] [thrd:GroupCoordinator]: GroupCoordinator/1: Sent ApiVersionRequest (v3, 78 bytes @ 0, CorrId 15)
2021.01.16 01:40:33.448431 [ 55 ] {} <Notice> StorageKafka (kafka): [rdk:REQTMOUT] [thrd:GroupCoordinator]: GroupCoordinator/1: Timed out ApiVersionRequest in flight (after 10010ms, timeout #0)
2021.01.16 01:40:33.448708 [ 55 ] {} <Warning> StorageKafka (kafka): [rdk:FAIL] [thrd:GroupCoordinator]: GroupCoordinator: kafka1:19092: ApiVersionRequest failed: Local: Timed out: probably due to broker version < 0.10 (see api.version.request configuration) (after 10010ms in state APIVERSION_QUERY)
2021.01.16 01:40:33.449032 [ 55 ] {} <Debug> StorageKafka (kafka): [rdk:FEATURE] [thrd:GroupCoordinator]: GroupCoordinator/1: Updated enabled protocol features -ApiVersion to MsgVer1,BrokerBalancedConsumer,ThrottleTime,Sasl,SaslHandshake,BrokerGroupCoordinator,LZ4,OffsetTime,MsgVer2,IdempotentProducer,ZSTD,SaslAuthReq
2021.01.16 01:40:33.449302 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:COORD] [thrd:main]: Group "duplicates_when_commit_failed": marking the coordinator (1) dead: Local: Broker transport failure: OffsetCommitRequest failed
2021.01.16 01:40:33.449315 [ 55 ] {} <Warning> StorageKafka (kafka): [rdk:REQTMOUT] [thrd:GroupCoordinator]: GroupCoordinator/1: Timed out 1 in-flight, 0 retry-queued, 0 out-queue, 0 partially-sent requests
2021.01.16 01:40:33.449409 [ 58 ] {} <Debug> StorageKafka (kafka): [rdk:SEND] [thrd:kafka1:19092/bootstrap]: kafka1:19092/1: Sent MetadataRequest (v4, 80 bytes @ 0, CorrId 42)
2021.01.16 01:40:33.449437 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPCOORD] [thrd:main]: Group "duplicates_when_commit_failed" changing coordinator 1 -> -1
2021.01.16 01:40:33.449588 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:COORDCLEAR] [thrd:main]: Group "duplicates_when_commit_failed" broker kafka1:19092/1 is no longer coordinator
2021.01.16 01:40:33.449752 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPSTATE] [thrd:main]: Group "duplicates_when_commit_failed" changed state wait-coord -> query-coord (join-state wait-unassign-call)
2021.01.16 01:40:33.449888 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPQUERY] [thrd:main]: kafka1:19092/1: Group "duplicates_when_commit_failed": querying for coordinator: OffsetCommitRequest failed
2021.01.16 01:40:33.450038 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPSTATE] [thrd:main]: Group "duplicates_when_commit_failed" changed state query-coord -> wait-coord (join-state wait-unassign-call)
2021.01.16 01:40:33.450114 [ 58 ] {} <Debug> StorageKafka (kafka): [rdk:SEND] [thrd:kafka1:19092/bootstrap]: kafka1:19092/1: Sent FindCoordinatorRequest (v2, 76 bytes @ 0, CorrId 43)
2021.01.16 01:40:33.450176 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:COMMIT] [thrd:main]: GroupCoordinator: OffsetCommit for 1 partition(s) in join-state wait-unassign-call: manual: returned: Local: Broker transport failure
2021.01.16 01:40:33.450316 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:COMMIT] [thrd:main]: Group "duplicates_when_commit_failed": unable to OffsetCommit in state wait-coord: Local: Broker transport failure: coordinator (none) is unavailable: retrying later
2021.01.16 01:40:43.934612 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPQUERY] [thrd:main]: kafka1:19092/1: Group "duplicates_when_commit_failed": querying for coordinator: OffsetCommitRequest failed
2021.01.16 01:40:43.934790 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:COMMIT] [thrd:main]: OffsetCommit for 1 partition(s) in join-state wait-unassign-call: manual: returned: Local: Waiting for coordinator
2021.01.16 01:40:43.934894 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:OFFSET] [thrd:main]: OffsetCommit failed for duplicates_when_commit_failed [0] at offset 21 in join-state wait-unassign-call: Local: Waiting for coordinator
2021.01.16 01:40:43.934934 [ 58 ] {} <Debug> StorageKafka (kafka): [rdk:SEND] [thrd:kafka1:19092/bootstrap]: kafka1:19092/1: Sent FindCoordinatorRequest (v2, 76 bytes @ 0, CorrId 44)
2021.01.16 01:40:43.935071 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:DUMP] [thrd:main]: Assignment dump (started_cnt=1, wait_stop_cnt=0)
2021.01.16 01:40:43.935157 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:DUMP_ALL] [thrd:main]: List with 1 partition(s):
2021.01.16 01:40:43.935241 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:DUMP_ALL] [thrd:main]:  duplicates_when_commit_failed [0] offset STORED
2021.01.16 01:40:43.935326 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:DUMP_PND] [thrd:main]: List with 0 partition(s):
2021.01.16 01:40:43.935396 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:DUMP_QRY] [thrd:main]: List with 0 partition(s):
2021.01.16 01:40:43.935422 [ 29 ] {} <Error> StorageKafka (kafka): Exception during commit attempt: Local: Waiting for coordinator
2021.01.16 01:40:43.935466 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:DUMP_REM] [thrd:main]: List with 0 partition(s):
2021.01.16 01:40:43.935562 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:ASSIGNDONE] [thrd:main]: Group "duplicates_when_commit_failed": assignment operations done in join-state wait-unassign-call (rebalance rejoin=false)
2021.01.16 01:40:43.935653 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPTIMEOUT] [thrd:main]: Group "duplicates_when_commit_failed": timed out 1 op(s), 0 remain
2021.01.16 01:40:43.935754 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:CGRPOP] [thrd:main]: Group "duplicates_when_commit_failed" received op OFFSET_COMMIT in state wait-coord (join-state wait-unassign-call)
2021.01.16 01:40:43.935840 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:COMMIT] [thrd:main]: OffsetCommit internal error: Local: Group partition assignment lost
2021.01.16 01:40:43.935966 [ 56 ] {} <Debug> StorageKafka (kafka): [rdk:COMMIT] [thrd:main]: OffsetCommit for -1 partition(s) in join-state wait-unassign-call: manual: returned: Local: Group partition assignment lost
2021.01.16 01:40:43.936348 [ 10 ] {} <Trace> BaseDaemon: Received signal 6
2021.01.16 01:40:43.937360 [ 63 ] {} <Fatal> BaseDaemon: ########################################
2021.01.16 01:40:43.937548 [ 63 ] {} <Fatal> BaseDaemon: (version 21.1.1.5689, build id: D4E97207D34D81D9D96D33F98FFEE664F640697D) (from thread 56) (no query) Received signal Aborted (6)
2021.01.16 01:40:43.937648 [ 63 ] {} <Fatal> BaseDaemon: 
2021.01.16 01:40:43.937891 [ 63 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f7ce201818b 0x7f7ce1ff7859 0x89e97ae 0x15a27c24 0x15a50d91 0x15a252f5 0x15a26009 0x15ad92ec 0x15aea75b 0x15a2a424 0x15b4f9fc 0x89e3fad 0x7f7ce21d3609 0x7f7ce20f4293
2021.01.16 01:40:43.938148 [ 63 ] {} <Fatal> BaseDaemon: 5. raise @ 0x4618b in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.01.16 01:40:43.938282 [ 63 ] {} <Fatal> BaseDaemon: 6. abort @ 0x25859 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.01.16 01:40:44.004745 [ 63 ] {} <Fatal> BaseDaemon: 7. abort @ 0x89e97ae in /usr/bin/clickhouse
2021.01.16 01:40:44.053597 [ 63 ] {} <Fatal> BaseDaemon: 8. ./obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka.c:4355: ? @ 0x15a27c24 in /usr/bin/clickhouse
2021.01.16 01:40:44.104204 [ 63 ] {} <Fatal> BaseDaemon: 9. ./obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_cgrp.c:2824: ? @ 0x15a50d91 in /usr/bin/clickhouse
2021.01.16 01:40:44.152829 [ 63 ] {} <Fatal> BaseDaemon: 10. ./obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdatomic.h:86: rd_kafka_buf_callback @ 0x15a252f5 in /usr/bin/clickhouse
2021.01.16 01:40:44.201539 [ 63 ] {} <Fatal> BaseDaemon: 11. ./obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_buf.c:0: rd_kafka_buf_handle_op @ 0x15a26009 in /usr/bin/clickhouse
2021.01.16 01:40:44.250367 [ 63 ] {} <Fatal> BaseDaemon: 12. ./obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_op.c:0: rd_kafka_op_handle @ 0x15ad92ec in /usr/bin/clickhouse
2021.01.16 01:40:44.299207 [ 63 ] {} <Fatal> BaseDaemon: 13. ./obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka_queue.c:500: rd_kafka_q_serve @ 0x15aea75b in /usr/bin/clickhouse
2021.01.16 01:40:44.348605 [ 63 ] {} <Fatal> BaseDaemon: 14. ./obj-x86_64-linux-gnu/../contrib/librdkafka/src/rdkafka.c:2005: rd_kafka_thread_main @ 0x15a2a424 in /usr/bin/clickhouse
2021.01.16 01:40:44.397792 [ 63 ] {} <Fatal> BaseDaemon: 15. ./obj-x86_64-linux-gnu/../contrib/librdkafka/src/tinycthread.c:586: _thrd_wrapper_function @ 0x15b4f9fc in /usr/bin/clickhouse
2021.01.16 01:40:44.457518 [ 63 ] {} <Fatal> BaseDaemon: 16. __tsan_thread_start_func @ 0x89e3fad in /usr/bin/clickhouse
2021.01.16 01:40:44.457729 [ 63 ] {} <Fatal> BaseDaemon: 17. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
2021.01.16 01:40:44.457867 [ 63 ] {} <Fatal> BaseDaemon: 18. __clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.01.16 01:40:45.566231 [ 63 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: DE711F4CBCF29E50994968EDB4981AA8. There is no information about the reference checksum.
2021.01.16 01:40:55.488586 [ 1 ] {} <Fatal> Application: Child process was terminated by signal 6.
```

See https://github.com/ClickHouse/ClickHouse/pull/18671#issuecomment-761550168

**Comment by edenhill** [comment]
Thank you for this timely report, we were just about to release.
Fix is on the way.

**Comment by edenhill** [comment]
The fix for this is in v1.6.0-RC2, can you give it a spin in your tests?

**Comment by filimonov** [comment]
> The fix for this is in v1.6.0-RC2, can you give it a spin in your tests?

Kafka-related tests are green now. Thank you!

**Comment by edenhill** [comment]
Awesome, thanks!
